pkg/debug/target.go

```diff
@@ -0,0 +1,289 @@
+	)
+	switch backend {
+	case "native":
+		p, err = native.Launch(processArgs, wd, foreground)
+	case "lldb":
+		p, err = betterGdbserialLaunchError(gdbserial.LLDBLaunch(processArgs, wd, foreground))
+	case "rr":
+		p, _, err = gdbserial.RecordAndReplay(processArgs, wd, false)
+	case "default":
+		if runtime.GOOS == "darwin" {
+			p, err = betterGdbserialLaunchError(gdbserial.LLDBLaunch(processArgs, wd, foreground))
+			break
+		}
+		p, err = native.Launch(processArgs, wd, foreground)
+	default:
+		return nil, fmt.Errorf("unknown backend %q", backend)
+	}
+	if err != nil {
+		return nil, err
+	}
+	t, err := New(p, runtime.GOOS, runtime.GOARCH, debugInfoDirs)
+	if err != nil {
+		p.Detach(true)
+		return nil, err
+	}
+	return t, nil
+}
+
+// OpenCoreOrRecording takes a path and opens either an RR recording or a core file
+// depending on the backend selection.
+// If a "rr" is not specified, it will attempt to open a core file.
+// If opening a core file argv0 should be the path to the binary that produced the core file.
+// If debugInfoDirs is provided those directories will be included when looking up
+// debug information that is separate from the binary.
+func OpenCoreOrRecording(backend, path, argv0 string, debugInfoDirs []string) (*Target, error) {
+	var (
+		p   proc.Process
+		err error
+
+		os   = runtime.GOOS
+		arch = runtime.GOARCH
+	)
+	switch backend {
+	case "rr":
+		p, err = gdbserial.Replay(path, false, false)
+	default:
+		p, os, arch, err = core.OpenCore(path, argv0)
+	}
+	if err != nil {
+		return nil, err
+	}
+	return New(p, os, arch, debugInfoDirs)
+}
+
+var errMacOSBackendUnavailable = errors.New("debugserver or lldb-server not found: install XCode's command line tools or lldb-server")
+
+func betterGdbserialLaunchError(p proc.Process, err error) (proc.Process, error) {
+	if runtime.GOOS != "darwin" {
+		return p, err
+	}
+	if _, isUnavailable := err.(*gdbserial.ErrBackendUnavailable); !isUnavailable {
+		return p, err
+	}
+
+	return p, errMacOSBackendUnavailable
+}
+
+// Initialize performs any setup that must be taken after
+// we have successfully attached to the process we are going to debug.
+// This includes any post-startup initialization the process must perform,
+// as well as setting the default goroutine and creating some initial breakpoints
+// that are set by default to catch when the process crashes or panics.
+func (t *Target) Initialize() error {
+	entry, err := t.Process.EntryPoint()
+	if err != nil {
+		return err
+	}
+	if err := t.bi.AddImage(t.Process.ExecutablePath(), entry); err != nil {
+		return err
+	}
+
+	if err := t.Process.Initialize(); err != nil {
+		return err
+	}
+
+	g, _ := proc.GetG(t.CurrentThread())
+	t.SetSelectedGoroutine(g)
+
+	createUnrecoveredPanicBreakpoint(t)
+	createFatalThrowBreakpoint(t)
+	return nil
+}
+
+// Detach will force the target to stop tracing the process.
+// If kill is true the process will be killed during the detach.
+func (t *Target) Detach(kill bool) error {
```

Comment:
This method, Pid, SelectedGoroutine, Recorded, Restart, Direction, When, Checkpoint, etc. do nothing since embedding will automatically take care of them.

Reply:
Agreed, they are redundant and can be removed for the sake of the diff. I created them originally because I didn't simply embed `proc.Process` at first, that change came a bit later.